### PR TITLE
Gradient scaling proof of concept

### PR DESCRIPTION
**Pull request by tophmatthews**
The problem this PR is trying to solve is the extremely poor geometric aspect ratio solves that are sometimes required in thermo-mechanical simulations. In some cases, greater than 1:100 radial-to-axial geometric ratios are required in 2DRz symmetry. Unfortunately, the "moving" things or physics in these cases are dominated by the behavior in the radial direction. The large geometric aspect ratio has been accommodated by extreme refinement in the axial direction where really nothing is changing. This leads to large computational requirements due to highly refined meshes, and will only get worse as the radial direction is refined due to new physics implementation. More recently, Layered1D has seen success in cutting up the 2D problem into 1D slices. This approach will suffer difficulties with contact, especially frictional and glued. There has been inconsistencies in results as well, but it is unclear where the problem lies (i.e. I don't have enough info to dig down why the changes are present, but want to at least call out they are there).

Some of my other PR's that have tried to attack this have died on the vine, as they focused on individual physics #14125, #14120. In attempt to revive this, I tried to make geometric scaling as generic as possible. This PR implements gradient scaling within `MooseVariableLevel`, a fundamental level for which all other objects request `_grad_u`. This allows native handling of geometrically scaled geometries, and has been shown here to match exactly for typical diffusion problems in 1D, 2D, 2DRz, and 3D, with and without AD, and first and second order elements. I have implementations for tensor mechanics problems in 1D and 2D, but can't get the 2DRz problem to work only because I know I'm screwing up the transform somehow, which I could use help on, but it shouldn't be a show stopper.

This work has been helped by the residual scaling by @lindsayad, but I'm not sure it's essential.

In the case of pure isotropic diffusion with a single variable, geometric scaling will (I think) make the problem a bit tougher to solve, since it essentially turns it into a anisotropic diffusion. However, for tensor mechanics problems where axis are separated by variables, and individually scaled, this should allow for elements with perfect aspect ratio to be used to simulate real-world parts with terrible geometric aspect ratios.

Here's the true question from @bwspenc that stopped #14125: Does poor aspect ratio really matter? Everything I read says good aspect ratio is essential. Consequently, `ElementQualityChecker` checks for things like poor aspect ratio already. However, with automatic scaling, maybe the problems with poor aspect ratios will be taken care of natively? I've had a hard time really answering that, and could use support one way or the other to push this or not.

Here's what I want: a 10x10 mesh that captures the exact same variable fidelity as a 10x1000 mesh. In other words, I want to run a problem that has mesh refinement defined by the physics, not by element quality. I don't care how I get there, but this seems a decent solution that is up for discussion.

### CODE REVIEW

**Comment by tophmatthews** [comment]
ping @bwspenc @sapitts @novasr @acasagran...I could use some healthy debate around this topic

File: framework/include/problems/SubProblem.h [hunk 1]
```diff
@@ -642,6 +642,16 @@ class SubProblem : public Problem
    */
   bool haveADObjects() const { return _have_ad_objects; }
 
+  /*
+   * Method for returning scaling vector
+   */
+  std::vector<Real> getScalingVector() const { return _gradient_scaling_vector; }
```
Comment (tophmatthews) [inline framework/include/problems/SubProblem.h]: This should be RealVectorValue, but I can't get past compiling errors...

File: framework/src/variables/MooseVariableData.C [hunk 2]
```diff
@@ -739,15 +741,56 @@ MooseVariableData<OutputType>::computeValues()
...
+  {
+    scaleGrad(_grad_u);
+    if (_need_grad_old)
+      scaleGrad(_grad_u_old);
+    if (_need_grad_older)
+      scaleGrad(_grad_u_older);
+    if (_need_grad_previous_nl)
+      scaleGrad(_grad_u_previous_nl);
+    if (_need_grad_dot || _need_grad_dotdot)
+      mooseError("No support yet for gradient scaling with grad_dot");
```
Comment (tophmatthews) [inline framework/src/variables/MooseVariableData.C]: I'm not sure if gradient scaling screws up grad_dot, I just have this in here to be safe.

**Comment by tophmatthews** [comment]
Anyone?

**Comment by novasr** [comment]
Will read soon ;-)

**Comment by tophmatthews** [comment]
I added tests for scaling with power law creep. Also if I drop the x-axis scaling, then 2DRz works just fine. This is getting closer to actually working at this point, so some conversation would be nice instead of doing this in a vacuum.

**Comment by novasr** [comment]
I'm not sure what I think about all this... my instinct says it won't work in mechanics, but atm I can't express a quantitative reason, other than properly accounting for mesh displacement while doing this scaling seems problematic. You sum it up well in your last sentence, you say you "want mesh refinement based on physics, not element quality"; or dare I substitute geometry for quality.  That's like a "eat your cake and have it too" kind-of a thing. It may be possible, but someone with deep understanding of large displacement mechanics, our approach in-code, and the ramifications of your scaling proposition will have to weigh-in.

**Comment by tophmatthews** [comment]
> deep understanding of large displacement mechanics, our approach in-code, and the ramifications of your scaling proposition will have to weigh-in.

So I don't have any of those things, which is why I appreciate the conversation, and why I am asking for the conversation.

What I do have here is several tests that shows it does work for temperature, elastic mechanics, and creep mechanics, in 1D, 2D, and 2DRz in the axial direction. Since no one else is really replying, perhaps I'll throw it on some bison cases to really show if it works or not.

And what's wrong wanting cake?

**Comment by novasr** [comment]
In your tests, how much is the mesh displacing?

**Comment by tophmatthews** [comment]
> In your tests, how much is the mesh displacing?

I set BC's between 1e-3 to 1e-4.

**Comment by tophmatthews** [comment]
Other tests (not used here) showed good performance with eigenstrains too. I should probably add some.

**Comment by tophmatthews** [comment]
not sure what is going on with civet, it works locally for me...

**Comment by bwspenc** [comment]
I put the "DO NOT MERGE" tag on this because per our discussion today, I would like to see some evidence that this actually works before we merge additional code like this into the framework. If it actually helps, I'm all for this, but I'm pretty skeptical that it will.

**Comment by tophmatthews** [comment]
☹️ 

0/3 for PRs on this topic...I'll try and get a full blown example out there to show if it helps. I'm afraid that the point is still being missed on what this is trying to do, so I'll try and clarify up my arguments. Hopefully it won't take another plane ride to push this forward after all this work.

**Comment by bwspenc** [comment]
I totally get the point of what you're trying to do.  I'm also convinced that we're just moving the problem from one place to another.  Just show me an example where this helps, and I'll be all over this!

**Comment by tophmatthews** [comment]
not done yet...

**Pull request closed without merging**